src/java/org/apache/cassandra/db/compaction/SimpleShardTracker.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.compaction;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.format.SSTableWriter;
+
+/**
+ * A shard tracker that uses the provided tokens as a complete list of split points. The first token is typically
+ * the minimum token.
+ */
+class SimpleShardTracker implements ShardTracker
+{
+    private final Token[] sortedTokens;
+    private int index;
+    private Token currentEnd;
+
+    SimpleShardTracker(Token[] sortedTokens)
+    {
+        assert sortedTokens.length > 0;
+        assert sortedTokens[0].isMinimum();
+        this.sortedTokens = sortedTokens;
+        this.index = 0;
+        this.currentEnd = shardEnd();
+    }
+
+    @Override
+    public Token shardStart()
+    {
+        return sortedTokens[index];
+    }
+
+    @Nullable
+    @Override
+    public Token shardEnd()
+    {
+        return index + 1 < sortedTokens.length ? sortedTokens[index + 1] : null;
```

Comment:
Nit: Calculating on `shardEnd()` and using `currentEnd` on `advanceTo` feels a bit inconsistent (this could be called repeatedly by a user).

I'd rather return `currentEnd` here and have a different method to set it. Or drop `currentEnd` entirely and use `shardEnd()` in `advanceTo` as well.

Reply:
I added a new method and utilized `currentEnd` in `shardEnd()`